p2p/transport/webrtc/udpmux/mux.go

```diff
@@ -0,0 +1,309 @@
+		cancel:               cancel,
+		socket:               socket,
+		unknownUfragCallback: unknownUfragCallback,
+		storage:              newUdpMuxStorage(),
+	}
+
+	mux.wg.Add(1)
+	go func() {
+		defer mux.wg.Done()
+		mux.readLoop()
+	}()
+	return mux
+}
+
+// GetListenAddresses implements ice.UDPMux
+func (mux *udpMux) GetListenAddresses() []net.Addr {
+	return []net.Addr{mux.socket.LocalAddr()}
+}
+
+// GetConn implements ice.UDPMux
+// It creates a net.PacketConn for a given ufrag if an existing
+// one cannot be  found. We differentiate IPv4 and IPv6 addresses
+// as a remote is capable of being reachable through multiple different
+// UDP addresses of the same IP address family (eg. Server-reflexive addresses
+// and peer-reflexive addresses).
+func (mux *udpMux) GetConn(ufrag string, addr net.Addr) (net.PacketConn, error) {
+	a, ok := addr.(*net.UDPAddr)
+	if !ok && addr != nil {
+		return nil, fmt.Errorf("unexpected address type: %T", addr)
+	}
+	isIPv6 := ok && a.IP.To4() == nil
+	return mux.getOrCreateConn(ufrag, isIPv6)
+}
+
+// Close implements ice.UDPMux
+func (mux *udpMux) Close() error {
+	select {
+	case <-mux.ctx.Done():
+		return nil
+	default:
+	}
+	mux.cancel()
+	mux.socket.Close()
+	mux.wg.Wait()
+	return nil
+}
+
+// RemoveConnByUfrag implements ice.UDPMux
+func (mux *udpMux) RemoveConnByUfrag(ufrag string) {
+	if ufrag != "" {
+		mux.storage.RemoveConnByUfrag(ufrag)
+	}
+}
+
+func (mux *udpMux) getOrCreateConn(ufrag string, isIPv6 bool) (net.PacketConn, error) {
+	select {
+	case <-mux.ctx.Done():
+		return nil, io.ErrClosedPipe
+	default:
+		return mux.storage.GetOrCreateConn(ufrag, isIPv6, mux)
+	}
+}
+
+// writeTo writes a packet to the underlying net.PacketConn
+func (mux *udpMux) writeTo(buf []byte, addr net.Addr) (int, error) {
+	return mux.socket.WriteTo(buf, addr)
+}
+
+func (mux *udpMux) readLoop() {
+	for {
+		select {
+		case <-mux.ctx.Done():
+			return
+		default:
+		}
+
+		buf := pool.Get(ReceiveMTU)
+
+		n, addr, err := mux.socket.ReadFrom(buf)
+		if err != nil {
+			log.Errorf("error reading from socket: %v", err)
+			pool.Put(buf)
+			if os.IsTimeout(err) {
+				continue
+			}
+			return
+		}
+		buf = buf[:n]
+
+		// a non-nil error signifies that the packet was not
+		// passed on to any connection, and therefore the current
+		// function has ownership of the packet. Otherwise, the
+		// ownership of the packet is passed to a connection
+		processErr := mux.processPacket(buf, addr)
+		if processErr != nil {
+			buf = buf[:cap(buf)]
+			pool.Put(buf)
+		}
+	}
+}
+
+func (mux *udpMux) processPacket(buf []byte, addr net.Addr) error {
+	udpAddr, ok := addr.(*net.UDPAddr)
+	if !ok {
+		return fmt.Errorf("underlying connection did not return a UDP address")
+	}
+	isIPv6 := udpAddr.IP.To4() == nil
+
+	// Connections are indexed by remote address. We firest
+	// check if the remote address has a connection associated
+	// with it. If yes, we push the received packet to the connection
+	// and loop again.
```

Comment:
What do you mean by "loop again"? There's not loop here.

Reply:
Comment not updated from a previous refactor, the loop is in the callee fn, will update it